packages/amplify-graphql-relational-transformer/src/schema.ts

```diff
@@ -120,179 +137,243 @@ export function ensureHasOneConnectionField(config: HasOneDirectiveConfiguration
   const conditionInput = ctx.output.getType(conditionInputName) as InputObjectTypeDefinitionNode;
   if (conditionInput) {
-    ctx.output.putType(updateFilterConnectionInputWithConnectionField(conditionInput, connectionAttributeName));
+    ctx.output.putType({
+      ...conditionInput,
+      fields: [
+        ...filterInput.fields!,
+        ...getFilterConnectionInputFieldsWithConnectionField([...conditionInput.fields!], connectionAttributeName),
+      ],
+    });
   }
   config.connectionFields.push(connectionAttributeName);
-}
+};
 
 /**
  * If the related type is a hasOne relationship, this creates a hasOne relation going the other way
  *    but using the same foreign key name as the hasOne model
  * If the related type is a hasMany relationship, this function sets the foreign key name to the name of the hasMany foreign key
  *    but does not add additional fields as this will be handled by the hasMany directive
  */
-export function ensureBelongsToConnectionField(config: BelongsToDirectiveConfiguration, ctx: TransformerContextProvider) {
+export const ensureBelongsToConnectionField = (config: BelongsToDirectiveConfiguration, ctx: TransformerContextProvider): void => {
   const { relationType, relatedType, relatedField } = config;
   if (relationType === 'hasOne') {
     ensureHasOneConnectionField(config, ctx);
   } else {
     // hasMany
     config.connectionFields.push(getConnectionAttributeName(relatedType.name.value, relatedField.name.value));
+    config.connectionFields.push(...getSortKeyFieldNames(relatedType));
   }
-}
+};
 
-export function ensureHasManyConnectionField(
+/**
+ * ensureHasManyConnectionField
+ */
+export const ensureHasManyConnectionField = (
   config: HasManyDirectiveConfiguration | ManyToManyDirectiveConfiguration,
   ctx: TransformerContextProvider,
-) {
-  const { field, fieldNodes, object, relatedType } = config;
+): void => {
+  const {
+    field, fieldNodes, object, relatedType,
+  } = config;
 
   // If fields were explicitly provided to the directive, there is nothing else to do here.
   if (fieldNodes.length > 0) {
     return;
   }
 
+  let connectionFieldName = 'id';
+  const sortKeyFields = getSortKeyFields(ctx, object);
+
+  object.fields!.forEach(objectField => {
+    objectField.directives!.forEach(directive => {
+      if (directive.name.value === 'primaryKey') {
+        connectionFieldName = objectField.name.value;
+      }
+    });
+  });
+
+  config.connectionFields.push(connectionFieldName, ...sortKeyFields.map(it => it.name.value));
+
   const connectionAttributeName = getConnectionAttributeName(object.name.value, field.name.value);
 
   const relatedTypeObject = ctx.output.getType(relatedType.name.value) as ObjectTypeDefinitionNode;
   if (relatedTypeObject) {
-    ctx.output.putType(updateTypeWithConnectionField(relatedTypeObject, connectionAttributeName, isNonNullType(field.type)));
+    const updatedFields = [...relatedTypeObject.fields!];
```

Comment:
Why do we need the bang here? Should we be defaulting to an empty list, or have a type that defines a required list?

Reply:
ObjectTypeDefinitionNode has fields optional, but since this is a related type it has to be an @model so it will always have fields or this can be rewritten to `relatedTypeObject.fields ?? []`